txjuju/_utils.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2016 Canonical Limited.  All rights reserved.
+
+import subprocess
+from collections import namedtuple
+
+
+class Executable(namedtuple("Executable", "filename envvars")):
+    """A single executable."""
+
+    def __new__(cls, filename, envvars=None):
+        """
+        @param filename: The path to the executable file.
+        @param envvars: The environment variables with which
+            to run the executable.
+        """
+        filename = str(filename) if filename else None
+        if envvars is not None:
+            if not hasattr(envvars, "items"):
+                envvars = dict(envvars)
+            envvars = {str(k): str(v) for k, v in envvars.items() if v}
+        return super(Executable, cls).__new__(cls, filename, envvars)
+
+    def __init__(self, *args, **kwargs):
+        if not self.filename:
+            raise ValueError("missing filename")
+
+    @property
+    def envvars(self):
+        """The environment variables used when running the executable."""
+        envvars = super(Executable, self).envvars
+        if envvars is None:
+            return None
```

Comment:
Why not return {} to keep the same type as the case where we have vars? It'd allow someone to dict.update(Executable.envvars) regardless of whether the executable had envvars settings. I guess this path isn't really covered in our use cases anyway because we source all of os.environ when creating the executable.